python/amici/conserved_moieties.py

```diff
@@ -0,0 +1,862 @@
+
+                matrix[k1] = []
+                matrix2[k1] = []
+                for i in range(num_species):
+                    if abs(column[i]) > MIN:
+                        matrix[k1].append(i)
+                        matrix2[k1].append(column[i])
+                ok = False
+                pivots[k1] = matrix[k1][0] if len(matrix[k1]) else MAX
+    K1 = sum(len(matrix[i]) > 0 for i in range(K))
+    return K == K1
+
+
+def _monte_carlo(
+        matched: Sequence[int],
+        J: Sequence[Sequence[int]],
+        J2: Sequence[Sequence[float]],
+        fields: Sequence[float],
+        int_matched: MutableSequence[int],
+        int_kernel_dim: int,
+        cls_species_idxs: MutableSequence[MutableSequence[int]],
+        cls_coefficients: MutableSequence[MutableSequence[float]],
+        num_species: int,
+        initial_temperature: float = 1,
+        cool_rate: float = 1e-3,
+        max_iter: int = 10
+) -> Tuple[bool, int, Sequence[int]]:
+    """MonteCarlo simulated annealing for finding integer MCLs
+
+    Finding integer solutions for the MCLs by Monte Carlo, see step (b) in
+    the De Martino (2014) paper and Eqs. 11-13 in the publication
+
+    :param matched:
+        number of found MCLs
+    :param J:
+        index of metabolites involved in a MCL
+    :param J2:
+        coefficients of metabolites involved in a MCL
+    :param fields:
+        actual number of metabolites involved in a MCL
+    :param int_matched:
+        actual matched MCLs
+    :param int_kernel_dim:
+        number of MCLs found in :math:`S`
+    :param cls_species_idxs:
+        Modified in-place.
+    :param cls_coefficients:
+        Modified in-place.
+    :param initial_temperature:
+        initial temperature
+    :param cool_rate:
+        cooling rate of simulated annealing
+    :param max_iter:
+        maximum number of MonteCarlo steps before changing to relaxation
+    :returns:
+        status of MC iteration, number of integer MCLs, number of MCLs,
+        metabolites and reaction indices, MCLs and integer MCLs as a tuple
+
+        status indicates if the currently found moiety by the Monte Carlo
+        process is linearly dependent (False) or linearly independent (True)
+        in case of linear dependence, the current Monte Carlo cycle can be
+        considered otherwise the algorithm retries Monte Carlo up to max_iter
+    """
+    dim = len(matched)
+    num = [int(2 * random.uniform(0, 1)) if len(J[i]) else 0
+           for i in range(dim)]
+    numtot = sum(num)
+
+    def compute_h():
+        H = 0
+        for i in range(dim):
+            H += fields[i] * num[i] ** 2
+            for j in range(len(J[i])):
+                H += J2[i][j] * num[i] * num[J[i][j]]
+        return H
+
+    H = compute_h()
+
+    MIN = 1e-9
+    count = 0
+    howmany = 0
+    T1 = initial_temperature
+    e = math.exp(-1 / T1)
+    while True:
+        en = int(random.uniform(0, 1) * dim)
```

Comment:
I think it would be great if the determination of conservation laws was deterministic. Use a fixed random seed here?